src/server/frontend_wayland/wl_data_device.cpp

```diff
@@ -125,8 +153,46 @@ void mf::WlDataDevice::set_selection(std::optional<wl_resource*> const& source,
     }
 }
 
+void mf::WlDataDevice::start_drag(
+    std::optional<wl_resource*> const& source,
+    wl_resource* origin,
+    std::optional<wl_resource*> const& icon,
+    uint32_t serial)
+{
+    // "The [origin surface] and client must have an active implicit grab that matches the serial"
+    if (!origin)
+    {
+        BOOST_THROW_EXCEPTION(
+            mw::ProtocolError(resource, Error::role, "Origin surface does not exist."));
+    }
+
+    validate_pointer_event(client->event_for(serial));
+
```

Comment:
We're missing a check that the serial corresponds to the initiation of an implicit-grab *on the origin surface*.

Reply:
Good catch